pytest_mitmproxy.py

```diff
@@ -1 +1,76 @@
 # -*- coding: utf-8 -*-
+import re
+import os
+import shutil
+import pytest
+from contextlib import contextmanager
+from subprocess import Popen, CalledProcessError, TimeoutExpired
+from tempfile import TemporaryFile
+from urllib.parse import urlparse
+
+# clean the dumps each run
+if os.path.isdir("traffic_dumps"):
+    shutil.rmtree("traffic_dumps")
+
+
+@contextmanager
+def new_proxy(name, scope, flt=None):
+    ''' Start a new mitmdump process and
+    return the address for clients to connect
+
+    The captured content will be written to
+    ./<dump_dir>/<scope>/<name>
+    '''
+    dump_dir = "traffic_dumps"
+
+    os.makedirs("{}/{}".format(dump_dir, scope), mode=0o755, exist_ok=True)
+    dump_file = os.path.join(dump_dir, scope, name)
+
+    args = ["mitmdump", "--port=0", "--wfile={}".format(dump_file)]
+    args.append(flt) if flt else None
+
+    def read(fp):
+        fp.seek(0)
+        return fp.read().decode("utf-8")
+
+    with TemporaryFile() as stdout, TemporaryFile() as stderr:
+        ps = Popen(
+            args,
+            stdout=stdout,
+            stderr=stderr
+        )
+        if ps.poll() is not None:
+            raise CalledProcessError(
+                ps.returncode,
+                ps.args,
+                output=read(stderr)
+            )
+        status = read(stdout)
+        url = urlparse(re.search(".*(http://[^\n]+)\n", status).group(1))
+        yield url
+
+        try:
+            ps.terminate()
+            retcode = ps.wait(timeout=5)
+        except TimeoutExpired:
+            ps.kill()
+            retcode = ps.wait()
+
+        if retcode != 0:
+            raise CalledProcessError(retcode, ps.args, output=read(stderr))
+
+
+@pytest.fixture
+def proxy(request):
+    mark = request.node.get_marker("dump_filter")
+    flt = mark.args[0] if mark else None
+    with new_proxy(request.node.name, request.scope, flt) as url:
+        yield url
+
+
+@pytest.fixture(scope="session")
+def session_proxy(request):
+    mark = request.node.get_marker("dump_filter")
+    flt = mark.args[0] if mark else None
```

Comment:
I don't think this is correct - session level fixtures looking at marks on individual tests? That doesn't work. Either: 

1. `request.node` isn't available (which I doubt, since I assume you've run this code)
2. `request.node` represents the **first** test only, and is never rerun for the next test
3. I'm taking crazy pills and this actually works?

Oh, and you should probably refactor these two lines into another function or a fixture:

```python
@pytest.fixture
 def dump_filter(request):
     mark = request.node.get_marker("dump_filter")
     return mark.args[0] if mark else None
```